File: composeApp/src/commonMain/kotlin/kr/hhp227/storygroup/ui/screens/group/CreateGroupViewModel.kt
package kr.hhp227.storygroup.ui.screens.group

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kr.hhp227.storygroup.shared.domain.model.Group
import kr.hhp227.storygroup.shared.domain.model.GroupJoinType
import kr.hhp227.storygroup.shared.domain.usecase.CreateGroupUseCase
import kr.hhp227.storygroup.shared.domain.usecase.UploadImageUseCase
import kr.hhp227.storygroup.ui.mvi.MviViewModel

/**
 * 그룹 만들기 — 이름/소개/가입방식+커버 이미지(CreatePost 첨부와 동일하게 선택 즉시 업로드).
 * 성공은 Event.Created 일회성 발화 — 화면이 세션 GroupsViewModel을 갱신하고 뒤로 복귀한다.
 * iosApp CreateGroupViewModel.swift와 1:1 미러
 */
class CreateGroupViewModel(
    private val createGroupUseCase: CreateGroupUseCase,
    private val uploadImageUseCase: UploadImageUseCase
) : ViewModel(), MviViewModel<CreateGroupViewModel.UiState, CreateGroupViewModel.Action, CreateGroupViewModel.Event> {
    private val _uiState = MutableStateFlow(UiState())
    override val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<Event>(extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
    override val event: Flow<Event> = _event.asSharedFlow()

    override fun onAction(action: Action) {
        when (action) {
            is Action.Submit -> submit(action.name, action.description, action.joinType)
            Action.ClearError -> _uiState.update { it.copy(error = null) }
            is Action.ChangeCoverImage -> changeCoverImage(action.bytes, action.fileName, action.contentType)
        }
    }

    private fun changeCoverImage(bytes: ByteArray, fileName: String, contentType: String) {
        if (_uiState.value.isUploadingImage) return

        _uiState.update { it.copy(isUploadingImage = true, error = null) }
        viewModelScope.launch {
            runCatching { uploadImageUseCase(bytes, fileName, contentType) }
                .onSuccess { url ->
                    _uiState.update { it.copy(isUploadingImage = false, image = url) }
                }
                .onFailure { e ->
                    _uiState.update { it.copy(isUploadingImage = false, error = e.message ?: "이미지 업로드에 실패했습니다.") }
                }
        }
    }

    private fun submit(name: String, description: String, joinType: GroupJoinType) {
        if (_uiState.value.isSaving) return
        if (name.isBlank()) {
            _uiState.update { it.copy(error = "그룹 이름을 입력해주세요.") }
            return
        }
        _uiState.update { it.copy(isSaving = true, error = null) }
        viewModelScope.launch {
            runCatching {
                createGroupUseCase(
                    name = name.trim(),
                    description = description.ifBlank { null },
                    image = _uiState.value.image,
                    joinType = joinType
                )
            }.onSuccess { group ->
                _uiState.update { it.copy(isSaving = false) }
                _event.tryEmit(Event.Created(group))
            }.onFailure { e ->
                _uiState.update { it.copy(isSaving = false, error = e.message ?: "그룹 생성에 실패했습니다.") }
            }
        }
    }

    data class UiState(
        val isSaving: Boolean = false,
        val error: String? = null,
        val image: String? = null,
        val isUploadingImage: Boolean = false
    )

    sealed interface Action {
        data class Submit(val name: String, val description: String, val joinType: GroupJoinType) : Action
        data object ClearError : Action
        class ChangeCoverImage(val bytes: ByteArray, val fileName: String, val contentType: String) : Action
    }

    sealed interface Event {
        data class Created(val group: Group) : Event
    }
}

File: iosApp/iosApp/UI/Screens/Group/CreateGroupViewModel.swift
import Combine
import Foundation
import Shared

/// 그룹 만들기 — composeApp CreateGroupViewModel.kt와 1:1 미러.
/// 이름/소개/가입방식+커버 이미지(CreatePost 첨부와 동일하게 선택 즉시 업로드).
/// 성공은 Event.created 일회성 발화 — 화면이 세션 GroupsViewModel을 갱신하고 닫는다.
final class CreateGroupViewModel: MviViewModel {
    @Published private(set) var uiState = UiState()

    let event = PassthroughSubject<Event, Never>()

    private let createGroupUseCase: CreateGroupUseCase

    private let uploadImageUseCase: UploadImageUseCase

    func onAction(_ action: Action) {
        switch action {
        case .submit(let name, let description, let joinType):
            submit(name: name, description: description, joinType: joinType)
        case .clearError: uiState.error = nil
        case .changeCoverImage(let data, let fileName, let contentType):
            changeCoverImage(data: data, fileName: fileName, contentType: contentType)
        }
    }

    private func changeCoverImage(data: Data, fileName: String, contentType: String) {
        if uiState.isUploadingImage { return }

        uiState.isUploadingImage = true
        uiState.error = nil
        Task { @MainActor in
            do {
                let url = try await uploadImageUseCase.invoke(
                    bytes: data.toKotlinByteArray(),
                    fileName: fileName,
                    contentType: contentType
                )
                uiState.isUploadingImage = false
                uiState.image = url
            } catch {
                uiState.isUploadingImage = false
                uiState.error = error.kotlinMessage(fallback: "이미지 업로드에 실패했습니다.")
            }
        }
    }

    private func submit(name: String, description: String, joinType: GroupJoinType) {
        if uiState.isSaving { return }

        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmedName.isEmpty {
            uiState.error = "그룹 이름을 입력해주세요."
            return
        }
        uiState.isSaving = true
        uiState.error = nil
        Task { @MainActor in
            do {
                let group = try await createGroupUseCase.invoke(
                    name: trimmedName,
                    description: description.isEmpty ? nil : description,
                    image: uiState.image,
                    joinType: joinType
                )
                uiState.isSaving = false
                event.send(.created(group))
            } catch {
                uiState.isSaving = false
                uiState.error = error.kotlinMessage(fallback: "그룹 생성에 실패했습니다.")
            }
        }
    }

    init(container: AppContainer) {
        createGroupUseCase = container.createGroupUseCase
        uploadImageUseCase = container.uploadImageUseCase
    }

    struct UiState {
        var isSaving = false
        var error: String? = nil
        var image: String? = nil
        var isUploadingImage = false
    }

    enum Action {
        case submit(name: String, description: String, joinType: GroupJoinType)
        case clearError
        case changeCoverImage(data: Data, fileName: String, contentType: String)
    }

    enum Event {
        case created(Group)
    }
}
